### ISSUE TITLE
Block-level DFA (--dfa-granularity block) — the literature recipe that actually works on transformers (F9d fallout)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#143]
F9d found our per-matmul DFA on MoE experts is COUNTERPRODUCTIVE — DFA-trained experts (held-out CE 6.89) are worse than the SAME experts left FROZEN at random init (6.64), which lose to BP (6.44). Gate-verified freeze. But the DFA-transformer literature says this is very likely our RECIPE, not DFA:

- Launay et al. 2020 (NeurIPS, "DFA Scales"): transformers have the largest DFA-BP gap; their working settings apply feedback at BLOCK scale (BP within block), or per-sublayer while KEEPING BP for embeddings + attention value matrices. "Common practices need rethinking" for transformers. Optimizer retuning narrowed the gap a lot.
- Oudjedi & Krzakala 2024 (PNAS, optical DFA): trained a 1.07B generative transformer to BP-LEVEL LOSS with DFA — projecting the output error by a random matrix to each decoder BLOCK (not each weight), last block on BP. Their frozen/shallow baseline performed WORSE than DFA — block-level DFA BEATS frozen, the opposite of our F9d.

The difference is GRANULARITY. We do the finest thing (vocab-dim error → fixed random B → each individual expert up/down matrix). The literature does coarse block-level feedback with BP inside the block. Ours is exactly the granularity Launay flagged.

Ask: `--dfa-granularity matmul|block` on franken-moe (default matmul = current behavior, byte-null):
- `block`: the fixed random feedback projects the output error to the MoE BLOCK's residual output (one B per block, ne=[d_model, vocab]); backprop THEN computes the individual expert up/down gradients WITHIN the block from that block-level error signal (standard BP local to the block, no weight transport across the block boundary). This is the Launay "macro" / Oudjedi decoder-block recipe.
- Provenance records dfa_granularity. Gate: block-DFA experts provably move (not frozen), and the block-level feedback is a distinct graph from matmul-DFA.

This is THE decisive experiment for whether the program's negative is a recipe artifact. If block-DFA beats frozen and narrows the BP gap at the F9c/F9d fixture, DFA-as-subordinate is revived (correctly). If block-DFA ALSO loses to frozen, the negative deepens. Higher priority than the F10/F11/F12 line — it directly addresses whether the whole program mismeasured DFA. Tao will draft the block-DFA re-run of F9c/F9d against this.

**Comment by OriPekelman** [comment]
Delivered in 9b6ffbe — `--dfa-granularity matmul|block` on `toy train franken-moe`, default `matmul` (byte-null).

## Mechanism

The routed sub-block is detached from the CE graph in **both** directions: input detached so no gradient crosses the boundary downward, output detached so CE cannot reach the experts from above. Both detaches are forward-identity (vendor-patch 0011), so the prediction is bit-unchanged — only the backward path is replaced. A second loss root then carries the block-level signal:

```
L_blk = sum( routed ⊙ B_blk · detach(p − y) )
```

whose gradient at `routed` is exactly the random-projected output error, so autodiff derives the individual expert up/down gradients from it by ordinary BP local to the block. One `B` per block, `d_model × vocab`, drawn from the existing DfaB axes (`b_seed + 31`) — no weight transport across the boundary. Provenance records `dfa_granularity`.

## The constraint you need before designing the re-run

**Block-DFA is a DENSE-lane recipe.** `--routing top1` is rejected, fail-loud: those experts dispatch through `mul_mat_id`, which has no ggml backward (toy#110), so "BP inside the block" is not expressible there. If F9c/F9d ran top1, the block-DFA arm cannot be dropped into that configuration — you would need the matmul-DFA / frozen / BP arms re-run **dense** too, or the comparison is not apples-to-apples. Worth settling before you spend the runs.

Two other guards, both fail-loud: non-`dfa-experts` policies are rejected (chain is already full BP), as is `--freeze-experts` (block-DFA exists to train them).

One gap against Launay's actual working setting: his recipe keeps **BP for embeddings and attention value matrices** while applying feedback at block scale. Our `bp-spine` policy does that, but it is top1-only (`bp_spine = top1 && ...`), so it does not compose with `block`. What shipped is block-level feedback on the experts with attention still per-matmul DFA. If the block-DFA arm underperforms, that residual is the obvious next suspect and a cheap follow-up ticket — say the word and I'll wire a dense bp-spine.

Confirmed to compose (all differ from the block baseline): `--moe-latent`, `--moe-shared`, `--shape wide`, `--optimizer muon`, `--attn-gate`.

## What the gate asserts — and what it deliberately does not

Asserted: block-DFA experts **provably move** (`experts_sig` start ≠ end, against the frozen control), the block curve **differs from matmul-DFA** so it is a distinct graph rather than a rename, plus determinism, flag-null, provenance, and the three guard rejections.

**Not asserted: which arm wins.** I ran a four-way smoke (frozen / matmul-DFA / block-DFA / chain-BP, E=8, wide, 200 steps on the fineweb smoke pack) and I am **not** reporting an ordering from it, because it does not survive its own error bars: the four arms land within 0.15 nats of each other and the ordering **flipped between the 150- and 200-step horizons** on the same seeds. That is noise, and pinning it into a gate would pin noise. The decisive comparison is yours to run at the F9c/F9d fixture, corpus and horizon, with whatever seed count you consider adequate — the instrument is what this ticket owes you, not the verdict.

## Methodology note, self-inflicted

Same class of error as the `--freeze-experts` one I flagged on #141, caught the same way. My first composition sweep invoked the runner binary directly instead of through `bin/toy`; the runner reads **env vars, not argv**, so every flag was silently ignored and all five combinations returned an identical loss. It looked like a real "no effect" finding. The tell was that `--shape wide` cannot possibly leave the curve unchanged. Re-run through the CLI, all five differ, as reported above.

Relatedly, the gate caught a bug this ticket introduced before it shipped: an edit closed `if top1` early, leaving the aux-loss reducer `ones_t` unuploaded on top1 lanes. As a persistent input it read zeros in silence, collapsing `t_aux` to 0 and killing load balancing with no error — the starvation and floor legs failed and named it. Fixed, with a comment recording why the upload must stay under `top1`.

CUDA twins hand-mirrored. Full battery green: 53 gates, CUDA included.

**Comment by OriPekelman** [comment]
Closing: shipped in `9b6ffbe` — feat(franken-moe): --dfa-granularity block — the Launay/Oudjedi recipe (toy#143)

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
